pkg/rbac/sharedData.go

```diff
@@ -32,116 +39,164 @@ type resource struct {
-			shared.csErr = queryerr
-			shared.csResources = []resource{}
-			return shared.csResources, shared.csErr
+		// get all cluster-scoped resources and cache in shared.csResources
+		err := cache.shared.getClusterScopedResources(cache, ctx)
+		if err == nil {
+			klog.V(6).Info("Successfully retrieved cluster scoped resources!")
 		}
+		// get all namespaces in cluster and cache in shared.namespaces.
+		err = cache.shared.GetSharedNamespaces(cache, ctx)
+		if err == nil {
+			klog.V(6).Info("Successfully retrieved shared namespaces!")
+		}
+		// get all managed clustsers in cache
+		err = cache.shared.GetSharedManagedClusters(cache, ctx)
+		if err == nil {
+			klog.V(6).Info("Successfully retrieved managed clusters!")
+		}
+
+		return err
+
+	}
+
+}
 
-		if rows != nil {
-			defer rows.Close()
+func sharedCacheValid(shared *SharedData) bool {
 
-			for rows.Next() {
-				var kind string
-				var apigroup string
-				err := rows.Scan(&apigroup, &kind)
-				if err != nil {
-					klog.Errorf("Error %s retrieving rows for query:%s for apigroup %s and kind %s", err.Error(), query,
-						apigroup, kind)
-					continue
-				}
+	if (time.Now().Before(shared.csUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL) * time.Millisecond))) &&
+		(time.Now().Before(shared.nsUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL) * time.Millisecond))) &&
+		(time.Now().Before(shared.mcUpdatedAt.Add(time.Duration(config.Cfg.SharedCacheTTL) * time.Millisecond))) {
+
+		return true
+	}
+	return false
+}
 
-				shared.csResources = append(shared.csResources, resource{apigroup: apigroup, kind: kind})
+func (shared *SharedData) getClusterScopedResources(cache *Cache, ctx context.Context) error {
+
+	// lock to prevent checking more than one at a time and check if cluster scoped resources already in cache
+	shared.csLock.Lock()
+	defer shared.csLock.Unlock()
+	//clear previous cache
+	shared.csResources = nil
+	shared.csErr = nil
+	klog.V(6).Info("Querying database for cluster-scoped resources.")
+
+	// Building query to get cluster scoped resources
+	// Original query: "SELECT DISTINCT(data->>apigroup, data->>kind) FROM search.resources WHERE
+	// cluster='local-cluster' AND namespace=NULL"
+	schemaTable := goqu.S("search").Table("resources")
+	ds := goqu.From(schemaTable)
+	query, _, err := ds.SelectDistinct(goqu.COALESCE(goqu.L(`"data"->>'apigroup'`), "").As("apigroup"),
+		goqu.COALESCE(goqu.L(`"data"->>'kind_plural'`), "").As("kind")).
+		Where(goqu.L(`"cluster"::TEXT = 'local-cluster'`), goqu.L(`"data"->>'namespace'`).IsNull()).ToSQL()
+	if err != nil {
+		klog.Errorf("Error creating query [%s]. Error: [%+v]", query, err)
+		shared.csErr = err
+		shared.csResources = []resource{}
+		return shared.csErr
+	}
 
+	rows, queryerr := cache.pool.Query(ctx, query)
+	if queryerr != nil {
+		klog.Errorf("Error resolving query [%s]. Error: [%+v]", query, queryerr.Error())
+		shared.csErr = queryerr
+		shared.csResources = []resource{}
+		return shared.csErr
+	}
```

Comment:
`shared.csUpdatedAt` is not updated in these error cases. We'll need to discuss how we'll handle this case.

- If we set `shared.csUpdatedAt` we won't be able to recover from the error until the cache expires.
- If we don't set `shared.csUpdatedAt` we might generate too many requests if we keep failing every time we retry to get this data.

We'll need something in between where errors are cached for a shorter time period.

Reply:
After thinking about this, this should work best:
- Here, update the timestamp to record that we made the API call.
- Then, in `sharedCacheValid()` we can decide how long an error response cache is valid.

Reply:
Okay, as discussed, will note this as optimization step in next user story.